### ISSUE TITLE
Consuming/clearing items from the launchQueue

### ISSUE THREAD

**Issue by hybridherbst** [WICG/web-app-launch#92]
I'm running into an issue with using launchQueue and setConsumer - the launchQueue is not cleared on subsequent page refresh when items are consumed and there doesn't seem to be a way to persistently remove items from the queue.

Consider the following scenario: 

1. you have a "viewer app".
2. user double-clicks a file supported by the app; app is launched with one item in launch queue.
3. launch queue is processed, file is displayed.
4. user uses in-app UI to select another file to view.
5. user refreshes the page (F5)
6. unexpectedly, the old file displays again, because the app receives another launchQueue with items (despite them having been processed in an earlier launch).
7. as far as I can see there's no way to differentiate between "user has refreshed the page but doesn't want the launchQueue to be applied since it's already displaying another file" and "user has double-clicked the same item again and gets a new window instance with the same launch queue, which should be processed".

Suggestions appreciated - not sure if I'm missing something. I'd like to declare a launchQueue item as "used" and not have it show up again on page refresh.

EDIT: seems that the launchQueue also has no timestamp or similar for when the item was added to the queue (which would help to differentiate with files that have been dropped later etc.). So this is also not useful to differentiate if a launch queue item is actually expected to be processed or is "outdated" / "used in the past".

**Comment by fallaciousreasoning** [comment]
@hybridherbst this sounds like a browser specific bug - would you mind filing a bug against the specific browser (I imagine you're using Chromium?).

https://bugs.chromium.org/p/chromium/issues/entry

**Comment by hybridherbst** [comment]
@fallaciousreasoning as I wasn't able to find anything related to the desired behaviour when refreshing a window that had a launchQueue, I'm not certain that this is a bug. What is the intended behaviour and where can I read more about that?

**Comment by fallaciousreasoning** [comment]
I think that items in the launch queue should only be processed once. I don't know if this is specified anywhere but it was definitely my original intention (it may have changed, but I'd be surprised).

The spec seems to imply that only unconsumed LaunchParams should be passed to the Launch Consumer:
https://wicg.github.io/web-app-launch/#setconsumer-method

**Comment by alancutter** [comment]
This behaviour ought to be clarified in the spec.

The behaviour you're seeing in Chrome comes from:
```
  // Reloads have the last sent launch params re-sent as they may contain live
  // file handles that should persist across reloads.
  if (last_sent_queued_launch_params_ &&
      handle->GetReloadType() != content::ReloadType::NONE) {
    SendLaunchParams(*last_sent_queued_launch_params_, handle->GetURL());
    return;
  }
```
This is intended for the case where you open a file in the app via its file_handlers and then refresh the page. Without this the app would lose the file handle.
@evanstade

Your issue raises a very good point that the app may have changed which file it is working on in the meantime and the old LaunchParams has gone stale.

**Comment by evanstade** [comment]
> as far as I can see there's no way to differentiate between "user has refreshed the page but doesn't want the launchQueue to be applied since it's already displaying another file" and "user has double-clicked the same item again and gets a new window instance with the same launch queue, which should be processed".

There is of course case 3, which is what this behavior is intended to address: "user didn't start displaying another file, has refreshed the page but expects the same file to still be open"

Perhaps the app's action URL should include something to indicate "?src=fileHandler" and this should be stripped when the user takes some action such as opening a new file?

It also seems like the app already ought to be using `history.pushState()` when the new file is opened. If not, I'd give that a try. Chrome might be able to automatically handle the case where pushState has been called. Since I don't work in this area any more, +@dmurph

**Comment by dmurph** [comment]
Interesting. Are these file handles allowed to be saved in IndexedDB yet? @a-sully. If so - then I would imagine we should probably remove this behavior.

If not - I wonder if there is a way to differentiate the reload from the first run, so they can be used or filtered.

**Comment by alancutter** [comment]
Could the file handle be made available for the duration of the site session somehow rather than being lost on refresh?

MDN says:
> A page session lasts as long as the tab or the browser is open, and survives over page reloads and restores.

**Comment by hybridherbst** [comment]
@dmurph yes, file handles can be stored in IndexedDB, that's what I'm doing in the application here.

That being said, the permissions for that are less permissive than launchQueue - I haven't found a clear pattern in Chrome but basically launchQueue is always allowed to access the fileHandle ("must" allow as per the spec, which makes sense) while file handles in IndexedDB have some heuristic for when the browser denies permission (and requires a user prompt again).

Maybe a file handle that was passed in via launchqueue could have better permissions when being retrieved from IndexedDB?

(Haven't gotten to trying out the history.pushState workaround mentioned above - but that is not an ideal state I think)

**Comment by dmurph** [comment]
Transferable navigable? https://html.spec.whatwg.org/multipage/document-sequences.html#traversable-navigable

I think there are thoughts about allowing these file handles to stay alive longer, in which case we should probably remove the behavior of having them re-populate on reload. I agree with @evanstade that the launch url should probably include some sort of metadata to communicate this came from a launch, and then that can be modified right away to disambiguate with a reload.

There is the other edge case that if we DO remove this from the queue, and the action url is reloaded, what does the website do? They have to handle that too. So there is always going to be some sort of reload edge case here.

**Comment by a-sully** [comment]
I'm less familiar with the specifics of the launch queue, but I can chime in on the FileSystemHandle bits of this discussion

> That being said, the permissions for that are less permissive than launchQueue - I haven't found a clear pattern in Chrome but basically launchQueue is always allowed to access the fileHandle ("must" allow as per the spec, which makes sense) while file handles in IndexedDB have some heuristic for when the browser denies permission (and requires a user prompt again).

Correct, having a `FileSystemHandle` != having _permission_ to access said handle

Currently, Chrome clears all File System permission state when you close the last tab for the site. Refreshing the page is considered a navigation, so this wipes permission state. [File Handling initially grants "readwrite" permission](https://github.com/WICG/file-handling/blob/main/explainer.md#launch) to the file, but after that, it's treated just like any other `FileSystemHandle` that you could otherwise get from [a picker](https://wicg.github.io/file-system-access/#api-showopenfilepicker) or [drag-and-drop](https://wicg.github.io/file-system-access/#drag-and-drop).

Unsurprisingly, this is not a popular behavior https://github.com/WICG/file-system-access/issues/297. Though we're actively working on making these permissions more persistent, so stay tuned!

**Comment by hybridherbst** [comment]
Diving deeper into this.

I was able to `use history.pushState` to basically mark the launchQueue as "done" 👍 and move the handle into indexedDb instead. This solves the immediate problem but still leaves open questions, e.g.
- should I have to do this
- do I get the same elevated handle permissions (that launchQueue guarantees) if the handle is deserialized from indexedDb - I guess yes, but I'm not sure.

@a-sully thanks for the additional info! It doesn't quite match what I'm seeing (e.g. refresh does _not_ always wipe permissions; PWAs seem to handle permissions differently; history.state is always empty when any field in it is a handle); would it be better if I open a clarifying question in the file-system-access repo instead of continuing here?

@dmurph I'm not exactly sure I understand what traversable navigables are/how they relate, where can I read more about that? (I tried implementing back/forward but since the fileHandle doesn't seem to persist in the state not sure how to handle that)

**Comment by a-sully** [comment]
> @a-sully thanks for the additional info! It doesn't quite match what I'm seeing (e.g. refresh does _not_ always wipe permissions; PWAs seem to handle permissions differently; history.state is always empty when any field in it is a handle); would it be better if I open a clarifying question in the file-system-access repo instead of continuing here?

Out of curiosity, are there other open tabs for the same origin when you're trying this? Permissions are only cleared  when _the last tab for the origin_ is closed

**Comment by hybridherbst** [comment]
It might be that I get a bit confused because of the extra permissions that an item has when it was in the launchQueue even when the launch queue isn't processed (so a file handle from indexedDb does indeed seem to have guaranteed permissions when it's _also_ in the launch queue). I'll test this more!

**Comment by dmurph** [comment]
I was thinking that one thing that could help is to propose a change to the permission lifetime of this handle to all "traversable navigables". But - would like to avoid doing that if possible :)

**Comment by alancutter** [comment]
I'd be keen for the reload scenario to be catered to by something more specific to file handles, having this re-enqueue behaviour in launch_handler impacts all the other kinds of launches too e.g. share target and link capturing.

**Comment by evanstade** [comment]
>  having this re-enqueue behaviour in launch_handler impacts all the other kinds of launches too e.g. share target and link capturing

Is this not desirable? I tend to think that after you are done with the share action, the app should navigate to a different page, so reloading wouldn't re-enqueue the launch params. However if you refreshed before the share action was completed, you probably want the launch items to be enqueued again.

**Comment by dmurph** [comment]
This seems like it's likely to cause more confusion than help, as it's impossible for the developer to know whether this was a launch or a reload - the use would expect the UX to be different here. E.g. in an editor, you wouldn't expect the editor to reload the file on reload, you expect to resume your editing session (Where you may have made changes). Also not re-queueing is much simpler to think about when consuming the API (at least in my opinion).

I chatted with some T&S folks and this was the conclusion:
> Currently, for other one-time permission models, a reload or navigation to same origin doesn't cause permission to expire, maybe can we use the same kind of lifetime for FSA handles’ ACLs as well?

I believe if that happens, then we can remove this behavior without affecting use-cases?

**Comment by evanstade** [comment]
> E.g. in an editor, you wouldn't expect the editor to reload the file on reload, you expect to resume your editing session (Where you may have made changes). 

This is true, but changes are lost either way (with re-enqueue or no re-enqueue). If you had made changes and they weren't auto-saved, a good editor would have a `beforeunload` handler that pops up a dialog and warns you.

Making FSA permissions more persistent has been a long process. It's in progress, but doesn't work yet.

> we could be displaying something like "Reloading this page will re-open .txt, are you sure you want to do that?"

I think the POST warning is because POST can have side effects which are possibly not desirable. In this example, reopening the .txt file doesn't really have a bad side effect. (Also, this interstitial is a pretty awful crutch that it would be better to avoid repeating.)

**Comment by alancutter** [comment]
Semantically I'm not a fan of the current reload+reenqueue behaviour because a reload isn't an app launch, it's a navigation within the same site session. I think the semantic mismatch is a root cause of the original issue raised.

**Comment by davepermen** [comment]
@evanstade in my case it directly causes a side-effect (a file upload and processing). and now (gladly still in the local debug phase) it never stops adding the file again and again. 

there needs to be a way to clear items.

**Comment by dmurph** [comment]
Now that the permission for storing a handle in IDB has launched in m122, we should remove this behavior.

**Comment by dmurph** [comment]
https://developer.chrome.com/blog/persistent-permissions-for-the-file-system-access-api

**Comment by Javdu10** [comment]
> Semantically I'm not a fan of the current reload+reenqueue behaviour because a reload isn't an app launch, it's a navigation within the same site session. I think the semantic mismatch is a root cause of the original issue raised.

Agreed, discussion was focused on file handles, but reload is not relaunch and even less so a "launch"

**Comment by kettanaito** [comment]
Are there any updates regarding this? This issue makes the launch queue pretty pointless in real-world apps as it breaks user flows if they happen upon a page reload (which is a first-class action in PWAs).

@hybridherbst, please, how did you manage to mark the queue as consumed with `history.pushState()`? I'm trying pushing a new entry but that does not solve this behavior:

```ts
// Push an empty entry since I'm not navigating anywhere.
history.pushState(null, '')
```

Even if I use this for navigation, the launch queue keeps the previous file handle, breaking the app on subsequent reloads as the app gets tricked by the browser into thinking the user is requesting to open a file (when they aren't). 

Given persisted FileSystem API permissions, this behavior shouldn't be exhibited at all. The launch queue should be marked as consumed (conditions may apply, but there must be a way to explicitly mark it as consumed) and discard the file handle. And this isn't a handle identity issue either. It is not the file handle that marks the launch as consumed, _it is the developer_. You can very well trigger to open the same file multiple times throughout your workflow and it must be supported.

1. User opens file handle 1 via launch queue (from their system).
2. User closes the file.
3. User opens file handle 2 via the app itself.
4. User remembers they wanted to edit file 1.
5. User opens file handle 1 _via launch queue_ again.
6. The file is opened correctly.

All of these flows are solved if the `setConsumer` callback can decide when the launch has been handled on the individual launch attempts. 

## Specification vs relatity

I don't believe the paragraph 4.2.2 from the spec is the behavior that Chrome currently exhibits:

> Set queue's [unconsumed launch params](https://wicg.github.io/web-app-launch/#dfn-unconsumed-launch-params) to the empty list.
> —[spec](https://wicg.github.io/web-app-launch/#processing-unconsumed-launch-params)

The launch params are not set to an empty list, otherwise they wouldn't survive page reload. 

I find it troubling that the specification does not define who and when marks launch params as _consumed_. It the consumer method is present, it fires it off and (supposedly) resets the launch params. If it indeed behaved that way, that would've fixed the problems developers are having but introduce a set of different problems. 

## Proposal: explicitly consumed launch params

**Is it up to consideration to refine the spec to allow the developer to mark particular launch params as consumed/unconsumed?** I would highly recommend this route. This API can benefit greatly from being more declarative and more obvious to work with. 

```ts
window.launchQueue.setConsumer((params) => {
  if (predicate(params.files[0])) {
    // Mark this launch attempt as consumed.
    params.consume()
  }
})
```

- If `LaunchParams.consume()` method is invoked within the registered consumer, the launch params are treated as consumed. The launch params queue is cleared.
- If `LaunchParams.consume()` method is not invoked, then the launch params aren't treated as consumed and will survive the navigation (likely an intentional behavior as previously mentioned in the thread; keeping for backward compatibility).
- You consume or not consume the entire launch attempt, not individual file handles. If you want to process certain handles while leave other handles "for later", you should do that manually. This is an edge case and should be addressed as such (e.g. by persisting some file handles in IndexedDB and handling them later as you see fit).

The exact shape of this API is open to suggestions.

**Comment by kettanaito** [comment]
An even better solution to this would be to adhere to the existing spec and always clear the unconsumed params queue after the consumer is fired. It doesn't matter whether the consumer handled the launch attempt in any way. The consumer must be called _only_ as a result of user interaction (user opening a file in their file system) and never again. This eliminates the need of `.consume()` or any other changes to the existing API.

**Comment by kettanaito** [comment]
> Although, it looks like technically this can all be handled by having a custom url argument on the file handler case that is removed, and then detected later to detect reload. So - perhaps we don't need to spend our extremely limited cycles on this API if it works as-is.

This would've been possible if not for two things:

1. `LaunchParams.targetURL` is always an empty string when opening a file while the app session is already running (see #96).
2. Any search parameters present in `file_handlers[0].action` are _stripped away_. I suspect this is the browser's doing as I haven't found this as a step in the algo from the spec. 

As it is right now, the behavior remains broken (at least in Chromium) with no viable workaround at all.

**Comment by dmurph** [comment]
As a workaround, can you push state to ADD a query argument or fragment to signify that the file was received? That way, if the user reloads, you know the file was already processed?

Perhaps this should prioritize this: https://github.com/WICG/web-app-launch/issues/98

**Comment by kettanaito** [comment]
@dmurph, that won't work. By implementing your suggestion, we're introducing another bug—when the user opens a new file _after_ that query parameter has been set, now our launch handler will ignore the rightful request to open another file. We have to consider that launch consumer also fires for file handlers as well (see #97 for a potential bug in how that's implemented). 

The launch params persisting cannot be solved client-side without breaking things. The bright side here is that this is likely a bug in Chromium since the spec clearly states that the launch params MUST be cleared once they are consumed. Sadly, it doesn't define anywhere what "consumed" actually means, but neither does it say they should survive page reloads.

**Comment by dmurph** [comment]
Acknowledged, I think you are right. Two bugs:
-  [LaunchParams files are enqueued again on page navigation \[40204185\] - Visible to public - Chromium](https://crbug.com/40204185)
- [`LaunchParams.targetURL` is always an empty string when opening a file while the app is active. \[464314997\] - Visible to public - Chromium](https://crbug.com/464314997)

working on now

**Comment by dmurph** [comment]
I think there is an interesting thing here with having no launch queue queued at all on reload, and having it queued without the file.

I think, up till now, with navigation capturing at least, we likely always re-queue, but only if installed? this seems weird, devs have to support launchparams and non-launch params anyways, and it makes sense to only queue params when a launch happens, not a reload.

If we move to NavigationApi support, then that already tell dev it it's a reload - that is the long term goal.

so - in conclusion - I think we all want to just stop re-queueing anything on reload, is that right?

**Comment by kettanaito** [comment]
> I think we all want to just stop re-queueing anything on reload, is that right?

I do believe so, yes. As long as the file handlers function properly, too (reported [here](https://issues.chromium.org/issues/464314997)) since those can trigger the launch handler without the actual app launch happening (i.e. opening a file via your app when the app is already running). 

If I read the spec correctly, launch params must always be cleared once the launch handler has fired. Once that's the case, all the scenarios will function as expected (clear launch, reload, open when not running, open while running).

**Comment by dmurph** [comment]
Chromium change submitted, you can track the milestones here: https://chromiumdash.appspot.com/commit/8a6dab292d1a2d19fd15835860728b97434053eb (should be m146)
